ambry-network/src/main/java/com.github.ambry.network/LocalRequestResponseChannel.java

```diff
@@ -0,0 +1,161 @@
+      LocalChannelRequest localRequest = (LocalChannelRequest) request;
+      logger.debug("Added request for {}, queue size now {}", localRequest.processorId, requestQueue.size());
+    }
+  }
+
+  @Override
+  public Request receiveRequest() throws InterruptedException {
+    Request request = requestQueue.take();
+    if (request instanceof LocalChannelRequest) {
+      LocalChannelRequest localRequest = (LocalChannelRequest) request;
+      logger.debug("Removed request for {}, queue size now {}", localRequest.processorId, requestQueue.size());
+    }
+    return request;
+  }
+
+  @Override
+  public void sendResponse(Send payloadToSend, Request originalRequest, ServerNetworkResponseMetrics metrics) {
+    try {
+      LocalChannelRequest localRequest = (LocalChannelRequest) originalRequest;
+      ResponseInfo responseInfo =
+          new ResponseInfo(localRequest.requestInfo, null, byteBufferFromPayload(payloadToSend));
+      BlockingQueue<ResponseInfo> responseQueue = getResponseQueue(localRequest.processorId);
+      responseQueue.put(responseInfo);
+      logger.debug("Added response for {}, size now {}", localRequest.processorId, responseQueue.size());
+    } catch (IOException | InterruptedException ex) {
+      logger.error("Could not extract response", ex);
+    }
+  }
+
+  /**
+   * Receive all queued responses corresponding to requests matching a processor id.
+   * @param processorId the processor id to match.
+   * @param pollTimeoutMs the poll timeout in msec.
+   * @return the applicable responses.
+   */
+  public List<ResponseInfo> receiveResponses(int processorId, int pollTimeoutMs) {
+    BlockingQueue<ResponseInfo> responseQueue = getResponseQueue(processorId);
+    ResponseInfo firstResponse = null;
+    try {
+      firstResponse = responseQueue.poll(pollTimeoutMs, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ie) {
+      logger.debug("Interrupted polling responses for {}", processorId);
+    }
+    if (firstResponse == null) {
+      return Collections.emptyList();
+    }
+    List<ResponseInfo> responseList = new ArrayList<>();
+    responseList.add(firstResponse);
+    responseQueue.drainTo(responseList);
+    logger.debug("Returning {} responses for {}", responseList.size(), processorId);
+    return responseList;
+  }
+
+  /**
+   * @return the response list corresponding to a processor id.
+   * @param processorId the processor id to match.
+   */
+  private BlockingQueue<ResponseInfo> getResponseQueue(int processorId) {
+    return responseMap.computeIfAbsent(processorId, p -> new LinkedBlockingQueue<>());
+  }
+
+  @Override
+  public void closeConnection(Request request) throws InterruptedException {
+  }
+
+  @Override
+  public void shutdown() {
+  }
+
+  /**
+   * Utility to extract a byte buffer from a {@link Send} object, skipping the size header.
+   * @param payload the payload whose bytes we want.
+   */
+  static ByteBuffer byteBufferFromPayload(Send payload) throws IOException {
```

Comment:
I guess this method is temporary before we find a better way to avoid the extra serialization and then deserialization?

Reply:
Yes, added a TODO.